Source/DafnyRuntime/DafnyRuntime.py

```diff
@@ -39,57 +45,159 @@ def _break(name):
 def _tail_call():
     raise TailCall()
 
-class Seq(list):
+class Seq(tuple):
     def __init__(self, __iterable = None, isStr = False):
         if __iterable is None:
             __iterable = []
-        self.isStr = isinstance(__iterable, str) or isStr
-        super().__init__(__iterable)
+        self.isStr = isStr or isinstance(__iterable, str) or (isinstance(__iterable, Seq) and __iterable.isStr) or (not isinstance(__iterable, GeneratorType) and all(isinstance(e, str) and len(e) == 1 for e in __iterable) and len(__iterable) > 0)
 
     @property
     def Elements(self):
         return self
 
     @property
     def UniqueElements(self):
-        return set(self)
+        return frozenset(self)
 
     def __repr__(self) -> str:
         if self.isStr:
-          return ''.join(self)
-        return super().__repr__()
+            return ''.join(self)
+        return '[' + ', '.join(map(repr, self)) + ']'
```

Comment:
Now that I look at this, `__repr__` should be an expression that recreates the Python `Seq` value, not something that looks roughly like the Dafny sequence display.

`__str__` can be whatever you like though, so that would be this version if desired.